src/Components/Components/src/Rendering/RenderTreeBuilder.cs

```diff
@@ -935,6 +952,54 @@ public void Dispose()
 
     internal Dictionary<string, int>? GetNamedEvents()
     {
```

Comment:
If I'm reading correctly, the algorithm here is a linear search and would run on every component render operation, so is equivalent to walking every rendered frame as many times as we render them.

If that's the case, would it also be practical to have the renderer wait until it's quiescent and is ready to invoke the submit handler, and *then* perform a linear search through everything in the current rendertrees to find the matching handler? That might be a good opportunity to take, as it would decouple this from the diffing system, and means we don't have to do any searches through content that later gets replaced during re-rendering.

Reply:
The other nice thing about that is that it could be done within `EndpointHtmlRenderer` only, and then it could use arbitrarily EndpointHtmlRenderer-specific logic to decide what to match. We can tie into as many HTML conventions as we like, for example avoiding the custom `@onsubmit:name` syntax and just defaulting to the form name or ID.

Reply:
This already only happens when we are trying to dispatch the form, so I don't feel super bad about it. The only reason is here is a stopgap before we have the compiler API in the Razor Compiler. Then this can go away.

Reply:
We won't need any special Razor API if we trigger this based on the name/id on the form. Seems like we can make this easier to use and more efficient.

Reply:
How would that help, we would be looking for a different attribute, wouldn't we? So, we would be having to do the linear search you mentioned, just on a different attribute.

Isn't valuable to capture the name when we are going through the render process independent of whether it is based on the name, the id, or any other attribute/directive?

As for the name/id. Isn't it problematic to attach additional semantics/meaning to concepts that already have a well defined meaning in HTML? There are legitimate reasons in HTML to give a name/id to a form, for example, to interact with it via JS. Why would that associate additional behavior with it?

Reply:
I've removed the bit where we scan the component list.

Reply:
> How would that help, we would be looking for a different attribute, wouldn't we? So, we would be having to do the linear search you mentioned, just on a different attribute.

My point was that by extracting the lookup from the diffing system, we'd be free to implement arbitrary patterns inside `EndpointHtmlRenderer` without needing any abstraction over the concept. We could choose to look for `name`/`id` even though they are HTML and possibly even `form`-specific patterns.

> As for the name/id. Isn't it problematic to attach additional semantics/meaning to concepts that already have a well defined meaning in HTML? There are legitimate reasons in HTML to give a name/id to a form, for example, to interact with it via JS. Why would that associate additional behavior with it?

Yeah, I take your point and think that's reasonable. I wasn't really meaning that adding a `name` would "associate additional behavior" but rather would suffice for making the form unique enough to identify when used in conjunction with `@onsubmit`. I'm not strictly against `@onsubmit:name` but was trying to think through how relevant and necessary it is to introduce a further syntactical element.

Overall I do recognize the value in being able to avoid a linear scan. I suspect if we want to, we could find ways to do a lookup without a linear scan with or without `@onsubmit:name`.

I was going to say that since your proposed addition here is a "stopgap" it would be fine but also removing it is fine too. When implementing https://github.com/dotnet/aspnetcore/issues/48760 I'll have to generalize how the named event tracking works (e.g., to support many-forms-one-handler) and so it's likely the stopgap implementation would change then anyway, so overall it probably doesn't matter either way right now.